### PR TITLE
Add route config validation

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Connects to Particular/V6Launch#68

Adds validation to route config:
 * When reading `MessageEndpointMappings` only apply routes to message types and only apply publisher config to event types
 * When configuring routes/publishers by type, throw if type is not a message/event
 * When configuring routes/publishers by assembly or namespace, only apply routes to message types and only apply publisher config to event types
 * When configuring routes/publishers by assembly or namespace, throw if assembly or namespace contains no message/event types

TODO:
 - [x] Update doco on best practice enforcement https://github.com/Particular/docs.particular.net/pull/1815

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@bording @andreasohlund @DavidBoike @timbussmann this PR breaks the best practice tests. I've checked in V5 and we had no tests for the best practices. The problem is, with this new logic we cannot configure publisher for a non-even type. This was also true for V5. So if somebody disables best practice enforcement, he's still not able to subscribe to a command. 

Is that a bug or a hidden feature?

**Comment by andreasohlund** [comment]
>Is that a bug or a hidden feature?

Just a coincidence I'd say

>he's still not able to subscribe to a command.

Can he still subscribe to a plain `IMessage` ?

**Comment by SzymonPobiega** [comment]
@andreasohlund no. And I believe, looking at the code here https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Unicast/Config/UnicastBus.cs#L182 that it was the same in V5.

**Comment by DavidBoike** [comment]
I am super OK with that, but probably bears mentioning in the 5to6 upgrade guide?

Seems like best practice enforcement is something that should tighten up on every major version release? But then again I'm not familiar enough with the situations where someone would disable those checks.

**Comment by SzymonPobiega** [comment]
According to the doco http://docs.particular.net/nservicebus/messaging/best-practice-enforcement (confirmed by me digging in the code) you can't disable best practice enforcement in V5. Overriding it is a new V6 feature. Question is: should we
 * Change the routing validation to somehow (no idea how) become more lax when users want to disable best practice validation?
 * Do not allow subscribing to non-events?

**Comment by andreasohlund** [comment]
>Do not allow subscribing to non-events?

Some users only use the "message conventions" so I'd say the only thing we can do it to prevent them from subscribing to commands and sending events.

That way you always have a escape hatch by , for the specific endpoint (usually tests), just omit the Command/Event conventions ?

**Comment by SzymonPobiega** [comment]
@andreasohlund conventions are taken into account so when I say

> Do not allow subscribing to non-events?

I mean do not allow subscribing to types that don't implement `IEvent` and don't match the defined event convention.

**Comment by andreasohlund** [comment]
>I mean do not allow subscribing to types that don't implement IEvent and don't match the defined event convention.

So far the ICommand/IEvent convention has been optional and some users only use the IMessage convention. 

Wouldn't this mean that they would be forced to use the IEvent convention? 
Is this something we actually want?

**Comment by SzymonPobiega** [comment]
Checked the V5 behavior inside https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Unicast/Config/UnicastBus.cs#L182 and it was different than my assumptions.

TL;DR V5 registers routes for subscribe messages even if the type is not an event type. That's enough for me to continue on this thing.

**Comment by SzymonPobiega** [comment]
@andreasohlund I think this is ready for another round of review now.

**Comment by andreasohlund** [comment]
Should we still consider merging the best practices into routing (where we think they belong) in the future? (raise an issue to track?)

File: src/NServiceBus.AcceptanceTests/Routing/When_sending_non_message_with_routing_configured_by_assembly.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                    catch (Exception ex)
+                    {
+                        c.Exception = ex;
+                        c.GotTheException = true;
+                    }
+                }))
+                .Done(c => c.GotTheException)
+                .Run();
+
+            StringAssert.Contains("No destination specified for messag", context.Exception.ToString());
```
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/When_sending_non_message_with_routing_configured_by_assembly.cs]: Typo

File: src/NServiceBus.AcceptanceTests/Routing/When_sending_non_message_with_routing_configured_via_mappings.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    using EndpointTemplates;
+    using NUnit.Framework;
+
+    public class When_sending_non_message_with_routing_configured_via_mappings : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task Should_throw_when_sending()
+        {
+            var context = await Scenario.Define<Context>()
+                .WithEndpoint<Receiver>()
```
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/When_sending_non_message_with_routing_configured_via_mappings.cs]: Do we even need the receiver endpoint, just send the message to your self? (or would we say that the test is more intuitive this way)

File: src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/AssemblyPublisherSourceTests.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class AssemblyPublisherSourceTests
+    {
+        [Test]
+        public void It_returns_only_event_types()
+        {
+            var source = new AssemblyPublisherSource(Assembly.GetExecutingAssembly(), PublisherAddress.CreateFromEndpointName("Destination"));
+            var routes = source.Generate(new Conventions()).ToArray();
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/AssemblyPublisherSourceTests.cs]: What would happen with a user that only uses the "message" convention? Can he publish? is there a need for an opt-in like we have for the autosubscribe?

File: src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/AssemblyPublisherSourceTests.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            var source = new AssemblyPublisherSource(typeof(string).Assembly, PublisherAddress.CreateFromEndpointName("Destination"));
+
+            Assert.That(() => source.Generate(new Conventions()).ToArray(), Throws.Exception.Message.Contains("Cannot configure publisher for assembly"));
+        }
+
+        class NonMessage
+        {
+        }
+
+        class NonEvent : IMessage
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/AssemblyPublisherSourceTests.cs]: Similar to my q above: If a user don't use the event convention wouldn't this block him?

File: src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/NamespacePublisherSourceTests.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            var source = new NamespacePublisherSource(Assembly.GetExecutingAssembly(), "NServiceBus.Core.Tests.Routing.NamespacePublisherSourceTest.NoMessages", PublisherAddress.CreateFromEndpointName("Destination"));
+
+            Assert.That(() => source.Generate(new Conventions()).ToArray(), Throws.Exception.Message.Contains("Cannot configure publisher for namespace"));
+        }
+    }
+}
+
+namespace NServiceBus.Core.Tests.Routing.NamespacePublisherSourceTest.OtherNamespace
+{
+    class ExcludedEvent : IEvent
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/NamespacePublisherSourceTests.cs]: So we only include types that are in the root of the namespace? 
Is that how v5 behaved? 
I would have assumed that if I include `Acme.Sales.Contracts` it would include all sub-namespaces if present?

File: src/NServiceBus.Core/Routing/AssemblyRouteSource.cs [hunk 6]
```diff
@@ -18,7 +19,17 @@ public AssemblyRouteSource(Assembly messageAssembly, UnicastRoute route)
...
         {
-            return messageAssembly.GetTypes().Where(t => conventions.IsMessageType(t)).Select(t => new RouteTableEntry(t, route));
+            var routes = messageAssembly.GetTypes()
+                .Where(t => conventions.IsMessageType(t))
+                .Select(t => new RouteTableEntry(t, route))
+                .ToArray();
+
+            if (!routes.Any())
+            {
+                throw new Exception($"Cannot configure routing for assembly {messageAssembly.GetName().Name} because it contains no types considered as messages. Message types have to either implement NServiceBus.IMessage interface or follow a defined message convention.");
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/AssemblyRouteSource.cs]: >follow

Change to "match" ?

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/AssemblyPublisherSource.cs [hunk 7]
```diff
@@ -18,7 +19,17 @@ public AssemblyPublisherSource(Assembly messageAssembly, PublisherAddress addres
...
         {
-            return messageAssembly.GetTypes().Where(conventions.IsEventType).Select(t => new PublisherTableEntry(t, address));
+            var entries = messageAssembly.GetTypes()
+                .Where(conventions.IsEventType)
+                .Select(t => new PublisherTableEntry(t, address))
+                .ToArray();
+
+            if (!entries.Any())
+            {
+                throw new Exception($"Cannot configure publisher for assembly {messageAssembly.GetName().Name} because it contains no types considered as events. Event types have to either implement NServiceBus.IEvent interface or follow a defined event convention.");
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/AssemblyPublisherSource.cs]: follow => match?

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/TypePublisherSource.cs [hunk 8]
```diff
@@ -17,6 +17,14 @@ public TypePublisherSource(Type messageType, PublisherAddress address)
 
         public IEnumerable<PublisherTableEntry> Generate(Conventions conventions)
         {
+            if (!conventions.IsMessageType(messageType))
+            {
+                throw new Exception($"Cannot configure publisher for type {messageType.FullName} because it is not considered a message. Message types have to either implement NServiceBus.IMessage interface or follow a defined message convention.");
+            }
+            if (!conventions.IsEventType(messageType))
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/TypePublisherSource.cs]: As I eluded to above I wonder if this should be more relaxed and be a `conventions.IsCommandType()` instead to not block users that only use the message convention?

File: src/NServiceBus.Core/Routing/TypeRouteSource.cs [hunk 9]
```diff
@@ -17,7 +17,12 @@ public TypeRouteSource(Type messageType, UnicastRoute route)
 
         public IEnumerable<RouteTableEntry> GenerateRoutes(Conventions conventions)
         {
+            if (!conventions.IsMessageType(messageType))
+            {
+                throw new Exception($"Cannot configure routing for type {messageType.FullName} because it is not considered a message. Message types have to either implement NServiceBus.IMessage interface or follow a defined message convention.");
+            }
             yield return new RouteTableEntry(messageType, route);
+
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/TypeRouteSource.cs]: :fire:

File: src/NServiceBus.AcceptanceTests/Routing/When_sending_non_message_with_routing_configured_via_mappings.cs (hunk 2 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_sending_non_message_with_routing_configured_via_mappings.cs]: We only need the type for the `AddMapping` to work but I think we don't need to start it.

File: src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/AssemblyPublisherSourceTests.cs (hunk 3 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/AssemblyPublisherSourceTests.cs]: No opt in. But this only applies to the code-first APIs. The MEMs work like they did. Should we allow non-events to be configured for publishing with the code-firs API?

File: src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/NamespacePublisherSourceTests.cs (hunk 5 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/NamespacePublisherSourceTests.cs]: Based on this https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Config/MessageEndpointMapping.cs#L143 I'd say it does not include subnamespaces.

Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenSubscriptions/NamespacePublisherSourceTests.cs]: That was a bit of a surprise to me but we need to be backwards compatible so :+1:

**Comment by SzymonPobiega** [comment]
@andreasohlund I pushed the changes. Now it will require a doco update.

**Comment by SzymonPobiega** [comment]
Also, the `AutoSubscribePlainMessages` thing has been removed in V6 with a message

> Encourages bad practices. IMessageSession.Subscribe should be explicitly used.

So automatic subscription is hard-wired to require types to be considered events. It that correct or should we auto-subscribe to non-commands when best practice enforcement is turned `off` ?

**Comment by andreasohlund** [comment]
>So automatic subscription is hard-wired to require types to be considered events. It that correct or should we auto-subscribe to non-commands when best practice enforcement is turned off ?

It seems reasonable to require the users to specify which messages are events to benefit from auto subscribe so I'd say we leave as is?

File: src/NServiceBus.AcceptanceTests/BestPractices/When_sending_events_bestpractices_disabled_on_endpoint.cs [hunk 10]
```diff
@@ -3,7 +3,6 @@
     using System.Threading.Tasks;
     using AcceptanceTesting;
     using EndpointTemplates;
-    using Features;
     using NUnit.Framework;
 
     public class When_sending_events_bestpractices_disabled_on_endpoint : NServiceBusAcceptanceTest
```
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_sending_events_bestpractices_disabled_on_endpoint.cs]: Perhaps shorten to: `When_sending_events_with_best_practices_disabled` ?

File: src/NServiceBus.AcceptanceTests/BestPractices/When_subscribing_to_command_bestpractices_disabled_on_endpoint.cs [hunk 11]
```diff
@@ -3,7 +3,6 @@
     using System.Threading.Tasks;
     using AcceptanceTesting;
     using EndpointTemplates;
-    using Features;
     using NUnit.Framework;
 
     public class When_subscribing_to_command_bestpractices_disabled_on_endpoint : NServiceBusAcceptanceTest
```
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_subscribing_to_command_bestpractices_disabled_on_endpoint.cs]: Perhaps shorten to: `When_subscribing_to_command_with_best_practices_disabled`

File: src/NServiceBus.AcceptanceTests/BestPractices/When_unsubscribing_to_command_bestpractices_disabled_on_endpoint.cs [hunk 12]
```diff
@@ -3,7 +3,6 @@
     using System.Threading.Tasks;
     using AcceptanceTesting;
     using EndpointTemplates;
-    using Features;
     using NUnit.Framework;
 
     public class When_unsubscribing_to_command_bestpractices_disabled_on_endpoint : NServiceBusAcceptanceTest
```
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_unsubscribing_to_command_bestpractices_disabled_on_endpoint.cs]: Perhaps shorten to: `When_unsubscribing_to_command_with_best_practices_disabled` ?

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt [hunk 13]
```diff
@@ -1693,10 +1694,6 @@ namespace NServiceBus.Features
     {
         protected internal override void Setup(NServiceBus.Features.FeatureConfigurationContext context) { }
     }
-    public class BestPracticeEnforcement : NServiceBus.Features.Feature
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: Needs a note in the upgrade guide and changes to http://docs.particular.net/nservicebus/messaging/best-practice-enforcement

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/AssemblyPublisherSource.cs [hunk 14]
```diff
@@ -16,9 +17,34 @@ public AssemblyPublisherSource(Assembly messageAssembly, PublisherAddress addres
...
+
+            if (!entries.Any())
+            {
+                throw new Exception($"Cannot configure publisher for assembly {messageAssembly.GetName().Name} because it contains no types considered as events. Event types have to either implement NServiceBus.IEvent interface or match a defined event convention.");
+            }
+
+            return entries;
+        }
+
+        public IEnumerable<PublisherTableEntry> GenerateWithouthBestPracticeEnforcement(Conventions conventions)
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/AssemblyPublisherSource.cs]: Typo: `Withouth`

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/IPublisherSource.cs [hunk 15]
```diff
@@ -5,7 +5,8 @@ namespace NServiceBus
 
     interface IPublisherSource
     {
-        IEnumerable<PublisherTableEntry> Generate(Conventions conventions);
+        IEnumerable<PublisherTableEntry> GenerateWithBestPracticeEnforcement(Conventions conventions);
+        IEnumerable<PublisherTableEntry> GenerateWithouthBestPracticeEnforcement(Conventions conventions);
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/IPublisherSource.cs]: same typo as above

File: src/NServiceBus.Core/Routing/MessagingBestPractices/BestPracticeEnforcement.cs [hunk 16]
```diff
@@ -1,50 +0,0 @@
-namespace NServiceBus.Features
-{
-    /// <summary>
-    /// Makes sure that messaging best practices are followed.
-    /// </summary>
-    public class BestPracticeEnforcement : Feature
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/MessagingBestPractices/BestPracticeEnforcement.cs]: RIP

File: src/NServiceBus.Core/Routing/RoutingFeature.cs [hunk 17]
```diff
@@ -87,5 +96,35 @@ protected internal override void Setup(FeatureConfigurationContext context)
                 }
             }
         }
+
+        void EnableBestPracticeEnforcement(FeatureConfigurationContext context)
+        {
+            var validations = new Validations(context.Settings.Get<Conventions>());
+
+            context.Pipeline.Register(
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/RoutingFeature.cs]: Now that routing checks best practises do we still need this? (Yes bus.Send("explicitAddress") would bypass but that API is meant to give users full control anyway)

File: src/NServiceBus.Core/Routing/RoutingFeature.cs [hunk 18]
```diff
@@ -87,5 +96,35 @@ protected internal override void Setup(FeatureConfigurationContext context)
...
+                "EnforceSendBestPractices",
+                new EnforceSendBestPracticesBehavior(validations),
+                "Enforces send messaging best practices");
+
+            context.Pipeline.Register(
+                "EnforceReplyBestPractices",
+                new EnforceReplyBestPracticesBehavior(validations),
+                "Enforces reply messaging best practices");
+
+            context.Pipeline.Register(
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/RoutingFeature.cs]: Now that routing checks best practices do we still need this?

File: src/NServiceBus.Core/Routing/RoutingFeature.cs [hunk 19]
```diff
@@ -87,5 +96,35 @@ protected internal override void Setup(FeatureConfigurationContext context)
...
+                "EnforcePublishBestPractices",
+                new EnforcePublishBestPracticesBehavior(validations),
+                "Enforces publish messaging best practices");
+
+            context.Pipeline.Register(
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/RoutingFeature.cs]: Now that routing checks best practises do we still need this?

File: src/NServiceBus.Core/Routing/RoutingFeature.cs [hunk 20]
```diff
@@ -87,5 +96,35 @@ protected internal override void Setup(FeatureConfigurationContext context)
...
+                "EnforceSubscribeBestPractices",
+                new EnforceSubscribeBestPracticesBehavior(validations),
+                "Enforces subscribe messaging best practices");
+
+            context.Pipeline.Register(
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/RoutingFeature.cs]: Now that routing checks best practices do we still need this?

File: src/NServiceBus.AcceptanceTests/BestPractices/When_sending_events_bestpractices_disabled_on_endpoint.cs (hunk 10 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_sending_events_bestpractices_disabled_on_endpoint.cs]: There is another tests with that name which checks variant with `SendOptions`.

File: src/NServiceBus.AcceptanceTests/BestPractices/When_subscribing_to_command_bestpractices_disabled_on_endpoint.cs (hunk 11 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_subscribing_to_command_bestpractices_disabled_on_endpoint.cs]: Same as above, there's another test with that name.

File: src/NServiceBus.AcceptanceTests/BestPractices/When_unsubscribing_to_command_bestpractices_disabled_on_endpoint.cs (hunk 12 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/BestPractices/When_unsubscribing_to_command_bestpractices_disabled_on_endpoint.cs]: Same here.

File: src/NServiceBus.Core/Routing/RoutingFeature.cs (hunk 18 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/RoutingFeature.cs]: Yes. The part in routing config applies only to subscriber side.

File: src/NServiceBus.Core/Routing/RoutingFeature.cs (hunk 19 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/RoutingFeature.cs]: Yes, the validation in routing config apply only to unicast transports which actually require subscriber-side configuration. On brokers you only call `Subscribe.

File: src/NServiceBus.Core/Routing/RoutingFeature.cs (hunk 20 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/RoutingFeature.cs]: Same as above.

**Comment by SzymonPobiega** [comment]
@andreasohlund @bording over to you.

**Comment by SzymonPobiega** [comment]
I think this is now good to go.

File: src/NServiceBus.Core/Routing/RoutingFeature.cs (hunk 18 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Routing/RoutingFeature.cs]: Yes but wouldn't routing be invoked when you call .Subscribe and throw even if this wasn't there?

File: src/NServiceBus.Core/Routing/TypeRouteSource.cs [hunk 21]
```diff
@@ -17,6 +17,10 @@ public TypeRouteSource(Type messageType, UnicastRoute route)
 
         public IEnumerable<RouteTableEntry> GenerateRoutes(Conventions conventions)
         {
+            if (!conventions.IsMessageType(messageType))
+            {
+                throw new Exception($"Cannot configure routing for type {messageType.FullName} because it is not considered a message. Message types have to either implement NServiceBus.IMessage interface or match a defined message convention.");
```
Comment (SimonCropp) [inline src/NServiceBus.Core/Routing/TypeRouteSource.cs]: i prefer single quotes around full type names

**Comment by SimonCropp** [comment]
apart from my preference of single quotes around things like namespaces and types in exception messages the rest looks good.

does this page need to be updated to reflect this http://docs.particular.net/nservicebus/messaging/messages-events-commands#validation-messages  ?

**Comment by SimonCropp** [comment]
:shipit:

**Comment by SzymonPobiega** [comment]
@SimonCropp I added the single quotes. 

As for the doco, these error messages are config-time (thrown from different place and in different context) so not sure if I should add them there. They are worth documenting though.

**Comment by SimonCropp** [comment]
yeah just nice to leverage some google fu foe when people dump the exception message into google

**Comment by SzymonPobiega** [comment]
@SimonCropp I've added that point to the WIP doco issue related to this PR.

File: src/NServiceBus.Core/Routing/MessageEndpointMappingCollection.cs [hunk 22]
```diff
@@ -175,6 +175,10 @@ internal void Apply(Publishers publishers, UnicastRoutingTable unicastRoutingTab
             {
                 m.Configure((type, endpointAddress) =>
                 {
+                    if (!conventions.IsMessageType(type))
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageEndpointMappingCollection.cs]: Why is this check needed? Isn't that already covered by the APIs calling this method?

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt [hunk 23]
```diff
@@ -972,6 +972,7 @@ namespace NServiceBus
     public class RoutingSettings : NServiceBus.Configuration.AdvanceExtensibility.ExposeSettings
     {
         public RoutingSettings(NServiceBus.Settings.SettingsHolder settings) { }
+        public void DisableBestPracticeEnforcement() { }
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: Should we align the name with `DoNotEnforceBestPractices` on send options?

File: src/NServiceBus.Core.Tests/Routing/NamespaceRouteSourceTests.cs [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+    [TestFixture]
+    public class NamespaceRouteSourceTests
+    {
+        [Test]
+        public void It_returns_only_message_types()
+        {
+            var source = new NamespaceRouteSource(Assembly.GetExecutingAssembly(), "NServiceBus.Core.Tests.Routing.NamespaceRouteSourceTest", UnicastRoute.CreateFromEndpointName("Destination"));
+            var routes = source.GenerateRoutes(new Conventions()).ToArray();
+
+            Assert.IsTrue(routes.Any(r => r.MessageType == typeof(Message)));
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Routing/NamespaceRouteSourceTests.cs]: I general I prefer Collection asserts since they provide usually better assert messages when the test fails than Assert.IsTrue / IsFalse on collections. @andreasohlund seems to agree

File: src/NServiceBus.Core.Tests/Routing/NamespaceRouteSourceTests.cs [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+            Assert.IsFalse(routes.Any(r => r.MessageType == typeof(NonMessage)));
+        }
+
+        [Test]
+        public void It_returns_only_types_from_specified_namespace()
+        {
+            var source = new NamespaceRouteSource(Assembly.GetExecutingAssembly(), "NServiceBus.Core.Tests.Routing.NamespaceRouteSourceTest", UnicastRoute.CreateFromEndpointName("Destination"));
+            var routes = source.GenerateRoutes(new Conventions()).ToArray();
+
+            Assert.IsTrue(routes.Any(r => r.MessageType == typeof(Message)));
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Routing/NamespaceRouteSourceTests.cs]: https://github.com/Particular/NServiceBus/pull/4090/files#r76582463

File: src/NServiceBus.Core.Tests/Routing/NamespaceRouteSourceTests.cs [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+            Assert.IsFalse(routes.Any(r => r.MessageType == typeof(ExcludedMessage)));
+        }
+
+        [Test]
+        public void It_matches_namespace_in_case_insensitive_way()
+        {
+            var source = new NamespaceRouteSource(Assembly.GetExecutingAssembly(), "NServiceBus.Core.Tests.Routing.NAMESPACErouteSOURCEtest", UnicastRoute.CreateFromEndpointName("Destination"));
+            var routes = source.GenerateRoutes(new Conventions()).ToArray();
+
+            Assert.IsTrue(routes.Any(r => r.MessageType == typeof(Message)));
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Routing/NamespaceRouteSourceTests.cs]: https://github.com/Particular/NServiceBus/pull/4090/files#r76582463

File: src/NServiceBus.Core/Routing/MessageEndpointMappingCollection.cs (hunk 22 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/MessageEndpointMappingCollection.cs]: What do you mean? The MEMs are read from the config file. The thing calling the lambda does not check the conventions.

**Comment by SzymonPobiega** [comment]
@danielmarbach changed the method name and switched to collection asserts. Over to you.

File: src/NServiceBus.Core/Routing/MessageEndpointMappingCollection.cs (hunk 22 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageEndpointMappingCollection.cs]: Ok. overlooked this. thanks for the clarification

**Comment by danielmarbach** [comment]
@andreasohlund have a last look but from my side :+1: good to go
